### PR TITLE
Docker Compose Worker Documentation and Examples

### PR DESCRIPTION
**Pull request by Thumbscrew**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [ ] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by dklimpel** [comment]
Does it make sense to add a link to this documentation in the official docs, e.g. there:
https://github.com/matrix-org/synapse/blob/cde8af9a495cbc7f3d0207e3f17c37eddaee34e1/docs/setup/installation.md?plain=1#L26

File: changelog.d/12737.doc [hunk 1]
```diff
@@ -0,0 +1 @@
+Add documentation for how to configure Synapse with Workers using Docker Compose. Includes example worker config and docker-compose.yaml.
```
Comment (richvdh) [inline changelog.d/12737.doc:1]:
```suggestion
Add documentation for how to configure Synapse with Workers using Docker Compose. Includes example worker config and docker-compose.yaml. Contributed by @Thumbscrew.
```

File: contrib/docker_compose_workers/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# Setting up Synapse with Workers using Docker Compose
+
+This section describes how deploy and manage Synapse and workers via [Docker Compose](https://docs.docker.com/compose/).
```
Comment (richvdh) [inline contrib/docker_compose_workers/README.md:3]:
```suggestion
This directory describes how deploy and manage Synapse and workers via [Docker Compose](https://docs.docker.com/compose/).
```

**Review by richvdh — COMMENTED** [review]
This looks great, thank you! A couple of suggestions below.

Please could you sign off your contribution following the instructions at https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off. You just need to add a comment to the PR with "Signed-off-by: Your Name <your@email.example.org>".

**Comment by richvdh** [comment]
> Does it make sense to add a link to this documentation in the official docs, e.g. there:

yes, I think that makes sense, but it might be easier once this PR has landed, so we can get the link right.

**Comment by Thumbscrew** [comment]
> This looks great, thank you! A couple of suggestions below.
> 
> Please could you sign off your contribution following the instructions at https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off. You just need to add a comment to the PR with "Signed-off-by: Your Name [your@email.example.org](mailto:your@email.example.org)".

No problem! I'd rather not put my full name on the PR but I'll follow the private sign off instructions.

**Comment by Thumbscrew** [comment]
Hi, I sent my sign off via email but haven't heard anything back. Please can you confirm if you received it?

Thanks

**Comment by DMRobertson** [comment]
> Hi, I sent my sign off via email but haven't heard anything back. Please can you confirm if you received it?
> 
> Thanks

Thanks, that's been received. @richvdh was there anything else outstanding you wanted to see here?

**Comment by DMRobertson** [comment]
> Hi, I sent my sign off via email but haven't heard anything back. Please can you confirm if you received it?
> 
> Thanks

Thanks, that's been received. @richvdh was there anything else outstanding you wanted to see here?

**Review by richvdh — APPROVED** [review]
lgtm. I'm going to assume the sytest failures are unrelated.

**Comment by dklimpel** [comment]
Sorry, it is a little bit late, but why have the federation sender a replication listener? IMO it is only needed for generic worker, when it is also a stream writer.

**Comment by DMRobertson** [comment]
We send out cache invalidation messages over replication. This allows worker A to change something in the database and tell worker B that a change has occured. Without this, worker B holds stale data in cache, which is often the cause of a gnarly bug.

**Comment by dklimpel** [comment]
Interesting neither worker doc nor https://github.com/matrix-org/synapse/blob/65e675504fe060e5e99e145be450fe4d492f404f/docker/configure_workers_and_start.py#L91-L96 
mention a replication listener for federation sender, in addition to http and redis replication.

**Comment by DMRobertson** [comment]
Hmm. Maybe I'm getting confused about the different kinds of replication we have again.

**Pull request merged**